alg/teca_regional_moisture_flux.cxx

```diff
@@ -0,0 +1,690 @@
+
+                return 0;
+                )
+                TECA_ERROR("Unsupported mask data type " << region_mask->get_class_name())
+                return -1;
+            )
+            TECA_ERROR("Unsupported IVT data type " << ivt_u->get_class_name())
+            return -1;
+        )
+        TECA_ERROR("Unsupported coordinate data type " << lon->get_class_name())
+        return -1;
+}
+}
+
+
+
+
+#if defined(TECA_HAS_CUDA)
+// CUDA codes
+namespace cuda_gpu
+{
+
+
+// **************************************************************************
+template <typename coord_t>
+__global__
+void transform_coordinates(coord_t *theta, coord_t *phi, coord_t *sin_phi,
+    const coord_t *lon, const coord_t *lat, unsigned long nlon, unsigned long nlat)
+{
+    coord_t rad_deg = M_PI / 180.0;
+
+    unsigned long q = teca_cuda_util::thread_id_to_array_index();
+
+    if (q < nlon)
+    {
+        theta[q] = lon[q] * rad_deg;
+    }
+
+    if (q < nlat)
+    {
+        coord_t phi_q = ( 90.0 - lat[q] ) * rad_deg;
+
+        phi[q] = phi_q;
+        sin_phi[q] = sin( phi_q );
+    }
+}
+
+// **************************************************************************
+template <typename coord_t, typename data_t>
+__global__
+void divergence(data_t *div, const data_t *ivt_u, const data_t *ivt_v,
+    const coord_t *theta, const coord_t *phi, const coord_t *sin_phi,
+    unsigned long nlon, unsigned long nlat)
+{
+    unsigned long q = teca_cuda_util::thread_id_to_array_index();
+
+    unsigned long i = q % nlon;
+    unsigned long j = q / nlon;
+
+    if ((i < 1) || (i >= nlon - 1) || (j < 1) || (j >= nlat - 1))
+        return;
+
+    // div = d / d theta ( ivt_theta )
+    //     + d / d phi ( ivt_phi sin(phi) )
+
+    // forward difference. using a centered difference here resulted in
+    // a larger overall error.
+    data_t dthe = theta[i + 1] - theta[i];
+    data_t dphi = phi[j + 1] - phi[j];
+
+    data_t divt_the = ivt_u[q + 1] - ivt_u[q];
+    data_t divtsin_phi =  ivt_v[q + nlon] * sin_phi[j + 1] - ivt_v[q] * sin_phi[j];
+
+    div[q] = divt_the / dthe + divtsin_phi / dphi;
+}
+
+// **************************************************************************
+template <typename coord_t, typename data_t>
+__global__
+void integrate(data_t *cell_int, const data_t *div,
+    const coord_t *theta, const coord_t *phi,
+    unsigned long nlon, unsigned long nlat)
+{
+    unsigned long q = teca_cuda_util::thread_id_to_array_index();
+
+    unsigned long i = q % nlon;
+    unsigned long j = q / nlon;
+
+    if ((i < 1) || (i >= nlon - 1) || (j < 1) || (j >= nlat - 1))
+        return;
+
+    // cell by cell compute:
+    //
+    // \iint div dA
+    //
+    // where:
+    //
+    // div = d / d theta ( ivt_theta )
+    //     + d / d phi ( ivt_phi sin(phi) )
+    //
+    // dA = d theta d phi
+    //
+
+    // area of the cell
+    data_t dthe = theta[i + 1] - theta[i];
+    data_t dphi = phi[j + 1] - phi[j];
+
+    data_t dA = dthe * dphi;
```

Comment:
shouldn't there be a multiplication by sin(phi[j]) here to account for convergence of the meridians (i.e., dA should go to 0 near the poles)?

Reply:
> shouldn't there be a multiplication by sin(phi[j]) here to account for convergence of the meridians (i.e., dA should go to 0 near the poles)?

explained above